drivers: counter: esp32: add driver support for general-purpose counters

glaubermaroto opened the pull request.

Commit:
esp32: drivers: counter: add support for general-purpose counters

Adds support for ESP32 general-purpose Counters

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

glaubermaroto commented:
Hi everyone. I've deleted a file that was inserted by mistake and was breaking the build.
Any comments here?

chrta commented on boards/xtensa/esp32/esp32_defconfig:
```diff
@@ -32,3 +32,8 @@ CONFIG_I2C_1=y
+CONFIG_HEAP_MEM_POOL_SIZE=98304
+CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE=4096
+CONFIG_POSIX_API=y
+CONFIG_THREAD_CUSTOM_DATA=y
```
Are all these setting necessary?

glaubermaroto replied:
Not all settings are necessary. I've kept `CONFIG_HEAP_MEM_POOL_SIZE=98304` - which is necessary to reference `k_calloc` - and removed the remaining.

glaubermaroto replied:
> Not all settings are necessary. I've kept `CONFIG_HEAP_MEM_POOL_SIZE=98304` - which is necessary to reference `k_calloc` - and removed the remaining.

@chrta  I have moved `CONFIG_HEAP_MEM_POOL_SIZE` from `esp32_dev_config` to both `tests/drivers/counter/counter_basic_api/boards/esp32.conf` and `samples/driver/counter/alarm/boards/esp32.conf` since it is a project specific configuration. The reserved heap was too much and was also updated.

sylvioalves commented:
@nordic-krch @andyross would you take sometime to review this?

andyross commented on drivers/counter/timer_esp32.c:
```diff
@@ -0,0 +1,290 @@
+
+static inline void counter_esp32_sem_take(const struct device *dev)
+{
+	k_sem_take(&DEV_DATA(dev)->sem, K_FOREVER);
```
You want a spinlock here, not a semaphore.  Semaphores are blocking IPC primitives.  They'll put the calling thread to sleep until the lock is released.  But I see below you're trying to take this same lock in an ISR, where clearly that isn't possible.  What will actually happen on contention is that the ISR will try to take the semaphore, fail to find a count to decrement, and then blow up with an assertion if you have that enabled (or otherwise fail spectacularly if you don't).

And it can't work even in principle: the thread that holds the lock might be the one you interrupted.

andyross reviewed (changes requested):
One issue with the choice of synchronization primitive needs to be fixed.  Beyond that the code looks very reasonable (but don't trust me for the dts and device model stuff!  That layer confuses me).

One bit on naming: I know the industry can't decide whether to use the word "timer" or "counter" or "clock" for the same family of devices, but in Zephyr a "timer" is something that interacts with the timing subsystem and implements the various z_clock_* APIs (yes, we aren't consistent either).  So adding a driver named "timer" to a directory named "counter" is maybe a little confusing.

nordic-krch commented on drivers/counter/timer_esp32.c:
```diff
@@ -0,0 +1,290 @@
+	timer_hal_intr_enable(&TIMX->hal);
+	timer_hal_set_alarm_enable(&TIMX->hal, TIMER_ALARM_EN);
+	DEV_DATA(dev)->callback = alarm_cfg->callback;
+	DEV_DATA(dev)->user_data = (struct counter_alarm_cfg *)alarm_cfg;
```
alarm_cfg is not persistent, may be created on stack. You probably want to store only user_data here.

glaubermaroto replied:
Good point. Actually, `user_data` was being used for alarm settings only. No need to keep `callback` and `user_data` split, I have included `alarm_cfg` for this purpose instead.

nordic-krch commented on drivers/counter/timer_esp32.c:
```diff
@@ -0,0 +1,290 @@
+
+static int counter_esp32_set_alarm(const struct device *dev, uint8_t chan_id,
+				   const struct counter_alarm_cfg *alarm_cfg)
+{
```
it's missing check against channel being busy.

glaubermaroto replied:
It is the intended behavior if I got the comment right. `chan_id` is not used in this timer implementation and is handled by `ARG_UNUSED` macro to avoid build-time warnings.

nordic-krch commented on drivers/counter/timer_esp32.c:
```diff
@@ -0,0 +1,290 @@
+	ARG_UNUSED(chan_id);
+
+	counter_esp32_sem_take(dev);
+	timer_hal_intr_disable(&TIMX->hal);
```
ideally you should clean `DEV_DATA(dev)->callback` which can be used to distinguish between channel being busy or not.

nordic-krch reviewed (commented):
i would recommend to rename PR and commit to `drivers: counter:...` as it implements counter API. Have you considered running `tests/drivers/counter/counter_basic_api`?

glaubermaroto commented:
@nordic-krch @andyross I have fixed some points as per your previous review. Please, take a look when convenient.

andyross reviewed (approved):
Synchronization looks great, thanks

glaubermaroto commented:
> i would recommend to rename PR and commit to `drivers: counter:...` as it implements counter API. Have you considered running `tests/drivers/counter/counter_basic_api`?

glaubermaroto commented:
> i would recommend to rename PR and commit to `drivers: counter:...` as it implements counter API. Have you considered running `tests/drivers/counter/counter_basic_api`?

Yes, I have run those tests. All passing.

glaubermaroto commented:
@nashif Please, would you take some time to review this?

Commit:
esp32: drivers: counter: update manifest to support counter driver

Update manifest to revision that supports counter driver

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

glaubermaroto commented:
@nashif I consider this PR complete. I have fixed what was pointed in the comments and updated `west.yml` to support the counter driver. Please, take a look when convenient.
